packages/react-server/core/renderMiddleware.js

```diff
@@ -105,16 +105,25 @@ module.exports = function(server, routes) {
 			navigateDfd.resolve();
 
 			if (err) {
-				logger.log("onNavigate received a non-2xx HTTP code", err);
-				handleResponseComplete(req, res, context, start, page);
-				if (err.status && err.status === 404) {
-					next();
-				} else if (err.status === 301 || err.status === 302 || err.status === 307) {
+				// The page can elect to proceed to render
+				// even with a non-2xx response.  If it
+				// _doesn't_ do so then we're done.
+				var done = !(page && page.getHaveDocument());
+
+				if (err.status === 301 || err.status === 302 || err.status === 307) {
 					res.redirect(err.status, err.redirectUrl);
-				} else {
-					next(err);
+				} else if (done) {
```

Comment:
> That means I'm responsible for actually putting the link in, but presumably if I'm claiming hasDocument on a redirect I know what I'm doing?

w/ the updated impl that moves `hasDocument` into the return object, I think I agree. It's weird (I have literally no idea how you'd implement that behavior w/ spring or raw java servlets), but as long as supporting it doesn't make the code more complicated, I think it seems reasonable. I think w/ the `setHaveDocument` approach, it'd be possible to intend for setHaveDocument to work for 500s or something, and accidentally have it apply to 503s as well. (Weird/strange edge case either way, I know. :) 
)